### PR TITLE
James 2525 blob store guice integration

### PR DESCRIPTION
**Pull request by jeantil**
This PR provides guice components that are required to actually use the swift blobstore in a James product. 

In more details, this provides a binding for `org.apache.james.blob.api.BlobStore` which is backed by `org.apache.james.blob.objectstorage.ObjectStorageBlobsDAO`. The binding reads the configuration to determine which swift auth protocol should be used, retrieve the corresponding identifiers and build the DAO.

This PR also provides the building blocks for integration testing of the product but does not modify the existing acceptance tests nor provide new ones.

### CODE REVIEW

**Comment by jeantil** [comment]
```
Stopped: vm://localhost#158
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.157 [main] DEBUG org.apache.activemq.util.ThreadPoolUtils - Shutdown of ExecutorService: java.util.concurrent.ThreadPoolExecutor@20411320[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0] is shutdown: true and terminated: true took: 0.000 seconds.
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.158 [main] DEBUG org.apache.activemq.util.ThreadPoolUtils - Forcing shutdown of ExecutorService: java.util.concurrent.ThreadPoolExecutor@25a5c7db[Running, pool size = 1, active threads = 0, queued tasks = 0, completed tasks = 6]
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.159 [main] DEBUG org.apache.activemq.transport.vm.VMTransportFactory - Shutting down VM connectors for broker: localhost
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.159 [main] INFO org.apache.activemq.broker.TransportConnector - Connector vm://localhost stopped
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.160 [main] INFO org.apache.james.queue.activemq.ActiveMQMailQueueBlobTest$MyFileSystem - No file specified
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.163 [main] INFO org.apache.activemq.broker.BrokerService - Apache ActiveMQ 5.15.5 (localhost, ID:583574a31b49-38103-1538755410454-0:1) is shutting down
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.164 [main] INFO org.apache.activemq.broker.TransportConnector - Connector tcp://localhost:61616 stopped
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.164 [ActiveMQ Transport Server Thread Handler: tcp://127.0.0.1:61616] INFO org.apache.activemq.transport.tcp.TcpTransportServer - socketQueue interrupted - stopping
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.164 [ActiveMQ Transport Server Thread Handler: tcp://127.0.0.1:61616] INFO org.apache.activemq.broker.TransportConnector - Could not accept connection during shutdown  : {}
[643f06620f7dd98d51d367e6a32d2b4ace783d49] java.lang.InterruptedException: null
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2014)
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2088)
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 	at java.util.concurrent.LinkedBlockingQueue.poll(LinkedBlockingQueue.java:467)
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 	at org.apache.activemq.transport.tcp.TcpTransportServer$1.run(TcpTransportServer.java:479)
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 	at java.lang.Thread.run(Thread.java:748)
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.173 [main] DEBUG org.apache.activemq.util.ThreadPoolUtils - Shutdown of ExecutorService: java.util.concurrent.ThreadPoolExecutor@7fa98a66[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 1499] is shutdown: true and terminated: true took: 0.001 seconds.
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.173 [main] DEBUG org.apache.activemq.util.ThreadPoolUtils - Forcing shutdown of ExecutorService: java.util.concurrent.ThreadPoolExecutor@7be3a9ce[Running, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.173 [main] INFO org.apache.activemq.broker.BrokerService - Apache ActiveMQ 5.15.5 (localhost, ID:583574a31b49-38103-1538755410454-0:1) uptime 55.752 seconds
[643f06620f7dd98d51d367e6a32d2b4ace783d49] 16:04:26.173 [main] INFO org.apache.activemq.broker.BrokerService - Apache ActiveMQ 5.15.5 (localhost, ID:583574a31b49-38103-1538755410454-0:1) is shutdown
[643f06620f7dd98d51d367e6a32d2b4ace783d49] [WARNING] Tests run: 85, Failures: 0, Errors: 0, Skipped: 5, Time elapsed: 55.756 s - in org.apache.james.queue.activemq.ActiveMQMailQueueBlobTest
```

**Comment by jeantil** [comment]
test this please

File: server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/ObjectStorageBlobsDAOProvider.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+public class ObjectStorageBlobsDAOProvider implements Provider<ObjectStorageBlobsDAO> {
+    private static final String OBJECTSTORAGE_CONFIGURATION_NAME = "objectstorage";
+
+    private static final String OBJECTSTORAGE_NAMESPACE = "objectstorage.namespace";
+    private static final String OBJECTSTORAGE_PROVIDER = "objectstorage.provider";
+    private static final String OBJECTSTORAGE_SWIFT_AUTH_API = "objectstorage.swift.authapi";
+
+    private final Configuration configuration;
+    private final BlobId.Factory blobIdFactory;
+    private final HashMap<String, Function<ContainerName, ObjectStorageBlobsDAO>> providersByName;
```
Comment (aduprat) [inline server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/ObjectStorageBlobsDAOProvider.java]: s/HashMap/ImmutableMap

File: server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/ObjectStorageBlobsDAOProvider.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    private final HashMap<String, Function<ContainerName, ObjectStorageBlobsDAO>> swiftAuthApiByName;
+
+    @Inject
+    public ObjectStorageBlobsDAOProvider(PropertiesProvider propertiesProvider,
+                                         BlobId.Factory blobIdFactory) throws ConfigurationException {
+        providersByName = new HashMap<>();
+        providersByName.put("swift", this::getSwiftObjectStorageBlobsDao);
+        swiftAuthApiByName = new HashMap<>();
+        providersByName.put("tempauth", this::getTempAuthBlobsDao);
```
Comment (aduprat) [inline server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/ObjectStorageBlobsDAOProvider.java]: s/providersByName/swiftAuthApiByName

Comment (jeantil) [reply server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/ObjectStorageBlobsDAOProvider.java]: 🙈

File: server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/ObjectStorageBlobsDAOProvider.java (hunk 1 above)
Comment (jeantil) [reply server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/ObjectStorageBlobsDAOProvider.java]: 🙈

File: server/container/guice/blob-api-guice/pom.xml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                        <goals>
+                            <goal>jar</goal>
+                        </goals>
+                        <configuration>
+                            <finalName>${project.artifactId}</finalName>
+                            <archive>
+                                <manifest>
+                                    <addClasspath>true</addClasspath>
+                                    <classpathPrefix>${project.artifactId}.lib/</classpathPrefix>
+                                    <mainClass>org.apache.james.CassandraJamesServerMain</mainClass>
```
Comment (mbaechler) [inline server/container/guice/blob-api-guice/pom.xml]: we don't need that, do we ?

File: server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/SwiftKeystone2ConfigurationReader.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        "objectstorage.swift.keystone2.tenantname";
+
+    public static SwiftKeystone2ObjectStorage.Configuration readSwiftConfiguration(Configuration configuration) {
+        String endpointStr = configuration.getString(OBJECTSTORAGE_SWIFT_ENDPOINT, null);
+        String crendentialsStr = configuration.getString(OBJECTSTORAGE_SWIFT_CREDENTIALS, null);
+        String userNameStr = configuration.getString(OBJECTSTORAGE_SWIFT_KEYSTONE_2_USERNAME, null);
+        String tenantNameStr = configuration.getString(OBJECTSTORAGE_SWIFT_KEYSTONE_2_TENANTNAME, null);
+
+        Preconditions.checkArgument(endpointStr != null,
+            OBJECTSTORAGE_SWIFT_ENDPOINT + "is a mandatory configuration value");
```
Comment (mbaechler) [inline server/container/guice/blob-objectstorage-guice/src/main/java/org/apache/james/modules/objectstorage/SwiftKeystone2ConfigurationReader.java]: you lack a space before `is`

File: server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftRule.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ ****************************************************************/
+
+package org.apache.james.blob.objectstorage;
+
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.testcontainers.containers.GenericContainer;
+
+public class DockerSwiftRule implements TestRule {
```
Comment (mbaechler) [inline server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftRule.java]: why would we need a Rule when we have extension ?

File: server/container/guice/blob-api-guice/pom.xml (hunk 3 above)
Comment (jeantil) [reply server/container/guice/blob-api-guice/pom.xml]: no we don't I used the cassandra module as a template for this module.

File: server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftRule.java (hunk 5 above)
Comment (jeantil) [reply server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftRule.java]: I followed the exact same design as the cassandra docker integration. it also has a rule and an extension. 

From my understanding the Rule is used to create a GuiceRule which is used for the CassandraStepDef

Comment (jeantil) [reply server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftRule.java]: see final usage here: https://github.com/jeantil/james-project/blob/JAMES-2525-blob-store-demo/server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CassandraStepdefs.java

File: server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftSingleton.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+public class DockerSwiftSingleton {
+
+    public static final DockerSwiftContainer singleton = new DockerSwiftContainer();
+
+    static {
+        singleton.start();
+    }
+
+    public static void restart() {
```
Comment (chibenwa) [inline server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftSingleton.java]: The restart might not be needed, it is a trick introduced for Cassandra testing...

File: server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftSingleton.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.objectstorage;
+
+public class DockerSwiftSingleton {
```
Comment (chibenwa) [inline server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/DockerSwiftSingleton.java]: How would you clean the state of the container accross tests?

File: server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java [hunk 8]
```diff
@@ -67,6 +69,22 @@ public static ObjectStorageBlobsDAOBuilder builder(SwiftKeystone3ObjectStorage.C
         return SwiftKeystone3ObjectStorage.daoBuilder(testConfig);
     }
 
+    public CompletableFuture<ContainerName> createContainer(ContainerName name) {
```
Comment (chibenwa) [inline server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java]: I have the impression that you synchronously execute the container creation then wrap it in a future.

Is thread sychronisation really needed here?

File: server/container/guice/blob-objectstorage-guice/src/test/java/org/apache/james/modules/objectstorage/guice/DockerSwiftRule.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+import org.junit.runners.model.Statement;
+import org.testcontainers.containers.GenericContainer;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.inject.Module;
+import com.google.inject.util.Modules;
+
+public class DockerSwiftRule implements GuiceModuleTestRule {
+
+    private org.apache.james.blob.objectstorage.DockerSwiftRule swiftContainer =
```
Comment (chibenwa) [inline server/container/guice/blob-objectstorage-guice/src/test/java/org/apache/james/modules/objectstorage/guice/DockerSwiftRule.java]: You can avoid these FQDN by renaming this class `DockerSwiftTestRule`

Comment (jeantil) [reply server/container/guice/blob-objectstorage-guice/src/test/java/org/apache/james/modules/objectstorage/guice/DockerSwiftRule.java]: I don't mind. Just for the record the same logic could be applied to the cassandra containers 
image

File: server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java (hunk 8 above)
Comment (jeantil) [reply server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java]: it is not, I have updated accordingly

**Comment by chibenwa** [comment]
test this please

**Comment by jeantil** [comment]
test this please

**Comment by jeantil** [comment]
test this please

again ...

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**